### PR TITLE
[CFRU Port] Simultaneous HP Reduction 

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Description
Simultaneous HP Reduction PR ported from CFRU. I've done a couple minor things differently like depracting some of the EWRAM variables. The cfru still used those and assigned the values back from the calculated values. I felt it is not needed since you can always fetch the data.

Special thanks to @Skeli789 for their CFRU implementation and doing the actual work

## Things to note in the release changelog:
```
gBattleMoveDamage, 
gMoveResultFlags
gIsCriticalHit 
```
Those above been deprecated. Instead the following should be used with an appropriate battler argument 
```
calculatedDamage[MAX_BATTLERS_COUNT];
calculatedCritChance[MAX_BATTLERS_COUNT];
moveResultFlags[MAX_BATTLERS_COUNT];
``` 
To have access in battle scripts callnatives can be used.
```
setmoveresultflags
clearmoveresultflags
jumpifmoveresultflags
jumpifcriticalhit
```

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
putting on draft. can't figure out why the test fails

**Comment by AlexOn1ine** [comment]
fixed the test

File: test/battle/spread_moves.c [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_HYPER_VOICE, playerLeft);
+        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, opponentRight);
+        MESSAGE("The opposing Wobbuffet is switched out with the Eject Button!");
+        MESSAGE("2 sent out Pikachu!");
+        ABILITY_POPUP(opponentLeft, ABILITY_EMERGENCY_EXIT);
+        MESSAGE("2 sent out Wynaut!");
+    }
+}
+
+DOUBLE_BATTLE_TEST("Spread Moves: No damage will be dealt to a mon in an invulnerable position - Surf")
```
Comment (AsparagusEduardo) [inline test/battle/spread_moves.c:31]: This should probably be Fly

File: test/battle/spread_moves.c [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    } WHEN {
+        TURN { MOVE(opponentLeft, MOVE_FLY, target: playerLeft); MOVE(playerLeft, MOVE_HYPER_VOICE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_HYPER_VOICE, playerLeft);
+        NOT HP_BAR(opponentLeft);
+        HP_BAR(opponentRight);
+    }
+}
+
+DOUBLE_BATTLE_TEST("Spread Moves: No damage will be dealt to a mon in an invulnerable position - Surf")
```
Comment (AsparagusEduardo) [inline test/battle/spread_moves.c:47]: Is this supposed to be for Dive?

File: test/battle/spread_moves.c [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        PLAYER(SPECIES_WYNAUT);
+        OPPONENT(SPECIES_ZAPDOS);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(playerLeft, MOVE_BUBBLE_BEAM, target: opponentLeft); MOVE(opponentLeft, MOVE_SURF); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_SURF, opponentLeft);
+    }
```
Comment (AsparagusEduardo) [inline test/battle/spread_moves.c:58]: What is this actually testing?

File: test/battle/spread_moves.c [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        TURN { MOVE(playerLeft, MOVE_EARTHQUAKE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_EARTHQUAKE, playerLeft);
+        ABILITY_POPUP(opponentLeft, ABILITY_DISGUISE);
+        ABILITY_POPUP(playerRight, ABILITY_ICE_FACE);
+        ABILITY_POPUP(opponentRight, ABILITY_ICE_FACE);
+    }
+}
+
+// Can be removed once the above test passes
```
Comment (AsparagusEduardo) [inline test/battle/spread_moves.c:236]: Will it be removed now xD

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
General recommendation:
I want to start promoting with the rest of the codebase that we start encapsulating whenever we can. With inline functions, we should be able to make our code more flexible.
Specifically, I'm thinking that we should add the following:
- `Get/SetBattleMoveDamage(battler)`
- `Get/SetBattleCritChance(battler)`
- `Get/SetBattleMoveResultFlags(battler)`

**Comment by AlexOn1ine** [comment]
> I want to start promoting with the rest of the codebase that we start encapsulating whenever we can. With inline functions, we should be able to make our code more flexible.

What's the motivation behind this generally and why would encapsulation make sense in this specific case?

**Comment by AsparagusEduardo** [comment]
> What's the motivation behind this generally and why would encapsulation make sense in this specific case?

This is something I want to talk more in depth on the server, so in this case might not be blocking for the purposes of the PR.
But basically, encapsulation *always* makes sense for data access and it's good practice in coding. Even if it's a single line, having new fields already encapsulated will save both us and users on the long run when we want to modify those methods to add new logic to them and even allow us to modify const data on runtime if needed.

**Comment by AlexOn1ine** [comment]
> > What's the motivation behind this generally and why would encapsulation make sense in this specific case?
> 
> This is something I want to talk more in depth on the server, so in this case might not be blocking for the purposes of the PR. But basically, encapsulation _always_ makes sense for data access and it's good practice in coding. Even if it's a single line, having new fields already encapsulated will save both us and users on the long run when we want to modify those methods to add new logic to them and even allow us to modify const data on runtime if needed.

Would love to see a discussion about it because I somewhat disagree with encapsulating for the sake of it. ~~Premature optimization is the route of all evil~~

**Comment by AlexOn1ine** [comment]
please supersede this pr. though I'm not going to support it with bugfixes

**Comment by AlexOn1ine** [comment]
I don't know if anyone will pick this up again and I don't want the work to get lost so I will address the review comments sometime next week and support this particular pr with bugfixes

**Comment by AsparagusEduardo** [comment]
> I don't know if anyone will pick this up again and I don't want the work to get lost so I will address the review comments sometime next week and support this particular pr with bugfixes

I was about to do so ^^;

**Comment by AsparagusEduardo** [comment]
#5770

**Pull request closed without merging**